### ISSUE TITLE
Callback for Training Data Evaluation for Nonbinary classifier

### ISSUE THREAD

**Issue by Skylion007** [ppwwyyxx/tensorpack#127]
I was wondering if there exist any good examples about how I could get metrics from the test dataset while training. (I couldn't really find any that were not classification based). This would really help to find the point at which a network begins to overfit for example. Obviously, I would start with the periodic callback, but I don't want to accidentally train on the dataset, and I might want to run specific metrics that aren't used in the training dataset. Would I have swap out the dataflows in the callback? I would really appreciate some guidance on how to best integrate this feature with having to overhall the entire class that I am using. Any help would be appreciated.

**Comment by ppwwyyxx** [comment]
What kind of metrics do you want?

**Comment by Skylion007** [comment]
Like lets say I have some ground truth data. Every 10th epoch or so, I would like to print out the cross_entropy between the predicted output and the expected output. This would be data that is different from the training data and thus would not be part of the dataflow. I am little confused about how to do this in a high level sense restarting the entire python process and loading the data from the last saved model.  I would really appreciate if I could get this metric so I knew when my model is overfitting. Maybe that could even be used as a halting condition of training phase?

**Comment by ppwwyyxx** [comment]
Almost every example prints something evaluated on the validation set. For example, in mnist-convnet.py, the following line:
```python
            InferenceRunner(    # run inference(for validation) after every epoch
                dataset_test,   # the DataFlow instance used for validation
                # Calculate both the cost and the error for this DataFlow
                [ScalarStats('cross_entropy_loss'), ClassificationError('incorrect')]),
```
Note that "cross_entropy_loss" must be an op you defined in the graph already.

The validation is run every epoch. You can use `PeriodicTrigger(InferenceRunner(.........), every_k_epochs=10)` to run it every 10 epoch.

**Comment by akshata1391** [comment]
For Dropout, if I set isTraining = True, then during inference of the validation set, is the dropout set to isTraining=False ?

**Comment by ppwwyyxx** [comment]
`Dropout` by default does this automatically. (use `is_training=None`)
If you set `is_training=True`, then even in inference time it is still True.

**Comment by akshata1391** [comment]
Thank you for the quick reply
